docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,750 @@
+`ndarray` subclass. Further, we propose some API changes like renaming or repurposing of keyword
+arguments and attributes. These changes are expected to break most of existing user codes. However,
+existing GT4Py functionality will remain or be extended and thus updating codebases to the new
+interface amounts mostly to updating attribute and keyword argument names.
+
+The removal of universal functions will require codes relying on ufuncs to be re-written by
+implementing these using third-party libraries.
+
+
+Detailed Description
+--------------------
+
+Functionality
+^^^^^^^^^^^^^
+
+We chose to use internally NumPy and CuPy `ndarrays` to store CPU and GPU buffers, respectively,
+since they are standard and reliable components of the Python ecosystem. We rely on those libraries
+to implement part of the functionality, like :code:`__setitem__` and :code:`__getitem__`, by
+forwarding the calls to the appropriate NumPy/CuPy function call. As a consequence, support for
+`dtypes` and other functionality is restricted to the common denominator of CuPy and NumPy.
+
+.. note:: In this document, references to NumPy and CuPy objects or functions use the standard
+    shorted prefiex for these libraries, that is :code:`np.` for NumPy and :code:`cp.` for CuPy.
+
+Data Interface
+^^^^^^^^^^^^^^
+
+Objects implementing the data interface have the attribute :code:`__gt_data_interface__`. This
+attribute is a dictionary which maps `device identifiers` to a dictionary similar to those
+defined as the :code:`__array_interface__` and the :code:`__cuda_array_interface__`.
+
+A device identifier can be one of:
+
++ :code:`None` denoting a buffer in main memory
++ :code:`"gpu"` denoting a buffer on a GPU.
+
+The mapped dictionary in turn must contain the following keys and mapped objects of pairs. Their
+meaning is the same as in the NumPy :code:`__array_interface__` and the
+:code:`__cuda_array_interface__`:
+
++ :code:`"shape": Tuple[int, ...]`
++ :code:`"typestr": str`
++ :code:`"data": Tuple[int, bool]`
++ :code:`"strides": Tuple[int, ...]`
+
+In Addition, the following optional keys can be contained:
+
++ :code:`"acquire": Optional[Callable[[], Any]]` Is called on all objects that are passed to a
+  stencil, before running computations. It can be used to trigger a copy to the respective device.
+  If the key is not in the dictionary or if the value is :code:`None`, no action is taken.
++ :code:`"dims": Optional[Sequence[str]]]` Specifies the semantic dimensions to which the
+  respective dimensions of the object correspond. Currently meaningful are :code:`"I"`,
+  :code:`"J"`, :code:`"K"`.
++ :code:`"halo": Optional[Sequence[Tuple[int, int]]]` A tuple of length ndim with entries which are
+  2-tuples of ints. Specifies the start and end boundary sizes on the respective dimension.
+  At stencil call time, this property is used to infer the compute domain.
+  :code:`"J"`, :code:`"K"`.
```

Comment:
Instead of specifying the halo, i.e. what is not included in the default computational domain of a stencil, I propose to directly specify the default domain. First, things outside of the default computational domain must not necessarily be a halo in the sense of something that is not "owned" by the current processor, it could also be a ghost cell or something we haven't thought about yet. Secondly it appears more natural to formulate and reason about overcomputing into the halo originating from the default domain. Instead of shrinking the halo one would extend the default domain, or paired with a per-stencil api as requested by @mcgibbon, would do something like:

```python
# some sort of domain object not yet designed, array of tuples like interface
default_domain = Domain((2, 2+nx), (2, 2+nx), (2, 2+ny))

# intitialize storages
out = gt.storage.storage.Storage.zeros_like(..., default_domain=default_domain)
in = gt.storage.storage.Storage.zeros(..., default_domain=default_domain)

# ...

# overcompute into the halo
laplacian(out, in, domain=default_domain.extend(1, 1)) # differs from current stencil api
laplacian(out, in)
```

Note that the domain concept itself needs further refinement (how does the domain of different fields, e.g. regular field <-> staggered field, relate to each other) and that it is desirable that overcomputing is handled without manual user intervention.

Reply:
> Instead of specifying the halo, i.e. what is not included in the default computational domain of a stencil, I propose to directly specify the default domain.

Yes, I totally agree with this. This is exactly the API we have for Quantity (you give it an origin and an extent for the compute domain). Within Quantity you'd use `default_domain=self.extent` to initialize a storage. Much nicer than deriving four halo sizes.

What we'd like to do when calling e.g. the laplacian is to use the default origin and extent from the first argument to the stencil, and indicate how to extend/contract that origin and extent using a keyword argument. The example here doesn't quite work because the default domain isn't global state, but something like `laplacian(out, in, domain=out.default_domain.extend(1, 1))` could work well.

If this sort of interface isn't provided by gt4py, we can already write our own wrapper on the existing API to do it.